fontbe/src/metrics_and_limits.rs

```diff
@@ -31,57 +35,178 @@ struct GlyphLimits {
-        self.advance_width_max = max(self.advance_width_max, advance);
-
         match glyph {
             Glyph::Simple(simple) => {
-                let points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
-                self.max_points = max(self.max_points, points);
-                self.max_contours = max(self.max_contours, simple.contours().len() as u16);
+                let num_points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
+                let num_contours = simple.contours().len() as u16;
+                self.max_points = max(self.max_points, num_points);
+                self.max_contours = max(self.max_contours, num_contours);
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        num_points,
+                        num_contours,
+                        ..Default::default()
+                    },
+                )
             }
             Glyph::Composite(composite) => {
                 let num_components = composite.components().len() as u16;
                 self.max_component_elements = max(self.max_component_elements, num_components);
+                let components = Some(composite.components().iter().map(|c| c.glyph).collect());
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        components,
+                        ..Default::default()
+                    },
+                )
             }
         };
     }
+
+    fn compute_component_limits(&self) -> ComponentLimits {
+        let mut depths = HashMap::<GlyphId, u16>::new();
+        let mut pending = self
+            .glyph_info
+            .iter()
+            .filter_map(|(gid, gi)| if gi.is_component() { Some(*gid) } else { None })
+            .collect::<Vec<_>>();
+        let mut components = Vec::with_capacity(8);
+        let mut max_points = 0;
+        let mut max_contours = 0;
+        while !pending.is_empty() {
+            let size_before = pending.len();
+            pending.retain(|gid| {
+                let glyph_info = self.glyph_info.get(gid).unwrap();
+                components.clear();
+                components.extend(
+                    glyph_info
+                        .components
+                        .as_ref()
+                        .unwrap()
+                        .iter()
+                        .map(|gid| self.glyph_info.get(gid).unwrap()),
+                );
+
+                // Update max points, contours while we're here
+                (max_points, max_contours) = components
+                    .iter()
+                    .filter_map(|c| {
+                        if !c.is_component() {
+                            Some((c.num_points, c.num_contours))
+                        } else {
+                            None
+                        }
```

Comment:
the problem is here I believe, with the filter_map you are skipping components that are in turn components and only folding the num_points/num_contours for the components that are simple glyphs.
Try to make a composite glyph that contains two components with different depths, i.e. one component points to a simple glyph, whereas the other one points to a component that in turn points to a simple glyph.
Your code would not consider the points/contours of the second nested component when computing the total number of points/contours for the composite as a whole

Reply:
I added a glyph called "c3" inside your NestedComponent.glyphs test file that has three components: "period" (simple glyph, depth=1), "c1" (in turn a composite referencing "period", depth=2) and "c2" (composite referencing "c1", depth=3). The expected num_points for "c3" as a whole should be 12, and the expected num_contours should be 3. It's max_depth should be 3. The maxp_with_deep_components test fails currently because I have adjusted the expectations.

Reply:
(facepalm) duh! Good catch, ty.

Reply:
Fixed but result still doesn't match. It is much closer, now the only diff is maxCompositePoints 103 vs 104, presumably due to more vigorous dropping of implicit oncurves in fontc. I'd like to land this and then wait for your fontmake update to get `glyf` to be identical to see if maxp then matches.